corehq/apps/data_interfaces/models.py

```diff
@@ -1214,6 +1319,101 @@ def to_dict(self):
         }
 
 
+class CaseDuplicateNew(models.Model):
+    id = models.BigAutoField(primary_key=True)
+    case_id = models.CharField(max_length=126, db_index=True)
+    action = models.ForeignKey("CaseDeduplicationActionDefinition", on_delete=models.CASCADE)
+    hash = models.CharField(max_length=256)
+
+    class Meta:
+        db_table = "data_interfaces_caseduplicate_new"
+        unique_together = ('case_id', 'action')
+        indexes = [
+            models.Index(fields=['hash', 'action_id'])
+        ]
+
+    def __str__(self):
+        return (
+            f"CaseDuplicateNew("
+            f"case_id={self.case_id}, action_id={self.action_id}, hash={self.hash})"
+        )
+
+    def delete(self, *args, check_for_orphans=True, **kwargs):
+        with transaction.atomic():
+            if check_for_orphans:
+                other_records = CaseDuplicateNew.objects.filter(
+                    action=self.action, hash=self.hash).exclude(case_id=self.case_id)[:2]
+
+                if other_records.count() == 1:
+                    # This will be orphaned when the current record is deleted, so delete it as well
+                    other_records[0].delete(check_for_orphans=False)
+
+            return super().delete(*args, **kwargs)
```

Comment:
Digging a bit deeper, the number of queries can be reduced by at least one, and probably more (see `Collector` source):

```suggestion
    def delete(self, using=None, keep_parents=False, check_for_orphans=True):
        if check_for_orphans:
            return self._delete_with_orphan_check(using, keep_parents)
        return super().delete(using=using, keep_parents=keep_parents)

    def _delete_with_orphan_check(self, using=None, keep_parents=False):
        # adapted from Model.delete
        using = using or router.db_for_write(self.__class__, instance=self)
        with transaction.atomic(using=using, savepoint=False):
            to_delete = [self]
            other_records = type(self).objects.filter(
                action=self.action, hash=self.hash).exclude(case_id=self.case_id)[:2]
            if len(other_records) == 1:
                # This will be orphaned when the current record is deleted, so delete it as well
                to_delete.append(other_records[0])

            for obj in to_delete:
                assert obj.pk is not None, (
                    "%s object can't be deleted because its %s attribute is set to None." %
                    (obj._meta.object_name, obj._meta.pk.attname)
                )

            collector = Collector(using=using)
            collector.collect(to_delete, keep_parents=keep_parents)
            return collector.delete()
```

This will require two additional imports above:
```py
from django.db import router
from django.db.models.deletion import Collector
```

Reply:
> Digging a bit deeper, the number of queries can be reduced by at least one, and probably more (see `Collector` source):

I appreciate the suggestion, but this feels like premature optimization to me. From the [django docs](https://docs.djangoproject.com/en/5.0/topics/db/transactions/#controlling-transactions-explicitly):

> You can disable the creation of savepoints for inner blocks by setting the savepoint argument to False... this option should only be used if the overhead of savepoints is noticeable. It has the drawback of breaking the error handling described above.

I'm open to changing this implementation in the future if the additional save point proves to be an issue, but I think keeping the code as is optimizes for readability at the moment (`savepoint`s and `Collector` are not well understood, and the implementation is more complicated/longer).

Reply:
Fair enough to revisit in the future if it becomes an issue.

It's unclear to me if your reference to the documentation was more to point out the caution against premature optimization, or more about the potential drawback of using `savepoint=False`.

I'm more concerned about reducing the number of database hints than eliminating the savepoint (although it's unclear if creating a savepoint adds yet another database hit). Notably, the `QuerySet.delete` implementation also uses [`savepoint=False`](https://github.com/django/django/blob/4787972c941b0d090cf083e84a98c1791bb2ae4b/django/db/models/deletion.py#L459), so it seems safe and  possibly preferred in the context of deletes.